src/container/builder_impl.rs

```diff
@@ -42,104 +43,167 @@ pub(super) struct ContainerBuilderImpl {
-                        utils::do_exec(&args[0], args, envs)?;
-
-                        if let Some(container) = &self.container {
-                            // the command / program is done executing
-                            container
-                                .refresh_state()?
-                                .update_status(ContainerStatus::Stopped)
-                                .save()?;
-                        }
-
-                        Ok(Process::Init(init))
-                    }
-                    Process::Parent(_) => unreachable!(),
-                }
+        // create the parent and child process structure so the parent and child process can sync with each other
+        let (mut parent, parent_channel) = parent::ParentProcess::new(self.rootless.clone())?;
+        let mut child = child::ChildProcess::new(parent_channel)?;
+
+        let cb = Box::new(|| {
+            if let Err(error) = container_init(
+                self.rootless.clone(),
+                self.spec.clone(),
+                self.syscall.clone(),
+                self.rootfs.clone(),
+                self.console_socket.clone(),
+                self.notify_path.clone(),
+                &mut child,
+            ) {
+                log::debug!("failed to run container_init: {:?}", error);
+                return -1;
             }
-            _ => unreachable!(),
+
+            0
+        });
+
+        let init_pid = fork::clone(cb, namespaces.clone_flags)?;
+        log::debug!("init pid is {:?}", init_pid);
+
+        parent.wait_for_child_ready(init_pid)?;
+
+        cmanager.add_task(init_pid)?;
+        if self.rootless.is_none() && linux.resources.is_some() && self.init {
+            cmanager.apply(&linux.resources.as_ref().unwrap())?;
+        }
+
+        // if file to write the pid to is specified, write pid of the child
+        if let Some(pid_file) = &self.pid_file {
+            fs::write(&pid_file, format!("{}", init_pid))?;
+        }
+
+        if let Some(container) = &self.container {
+            // update status and pid of the container process
+            container
+                .update_status(ContainerStatus::Created)
+                .set_creator(nix::unistd::geteuid().as_raw())
+                .set_pid(init_pid.as_raw())
+                .save()?;
+        }
+
+        Ok(())
+    }
+}
+
+fn container_init(
+    rootless: Option<Rootless>,
+    spec: Spec,
+    command: LinuxSyscall,
+    rootfs: PathBuf,
+    console_socket: Option<FileDescriptor>,
+    notify_name: PathBuf,
+    child: &mut child::ChildProcess,
+) -> Result<()> {
+    let linux = spec.linux.as_ref().unwrap();
+    let namespaces: Namespaces = linux.namespaces.clone().into();
+    // need to create the notify socket before we pivot root, since the unix
+    // domain socket used here is outside of the rootfs of container
+    let mut notify_socket: NotifyListener = NotifyListener::new(&notify_name)?;
```

Comment:
This is the line where it fails for me. Notify_name is the full path, which is longer than SUN_LEN.